# Probabilistic Reasoning: Bayes' Theorem

## 📚 Learning Objectives

By completing this notebook, you will:
- Apply Bayes' theorem to update probabilities
- Use prior, likelihood, and posterior in simple examples

## 🔗 Where this fits

**Builds on:** Unit 1, lesson 10 "Knowledge Representation and Reasoning" — reasoning with *certain* facts; Bayes is reasoning when the facts are uncertain.

**Used later in:** Course 02 (AIAT 112) — Unit 3, where Bayes' rule scales up to Bayesian networks and HMMs.

---


## 🎯 The question that 95 doctors out of 100 got wrong

In 1982 David Eddy put a question to practising American physicians. A woman has
a positive screening mammogram. The prevalence of breast cancer in the screened
population is **1%**; the test's sensitivity is **80%**; its false-positive rate
is **9.6%**. What is the probability she has cancer?

About **95 out of 100 physicians answered around 75%**. The correct answer is
**7.7%** — roughly one in thirteen. The doctors were not bad at arithmetic. They
were reading the test's accuracy as if it were the answer, and ignoring the fact
that in a population where only 1% are ill, the small false-positive rate applied
to the 99% who are healthy produces far more positive results than the disease does.

That mistake has a name — **base-rate neglect** — and it is not a curiosity. It
is repeated every time a screening programme is designed, every time a fraud
filter is deployed, and every time someone quotes a model's accuracy without
mentioning how rare the thing it detects is. The very first cell below computes
that exact calculation on numbers of the same shape, and prints **8.76%**.

### What goes wrong without Bayes' theorem

You confuse two different questions: *"how often does the test fire when the
disease is present?"* (sensitivity, 95% below) and *"how often is the disease
present when the test fires?"* (the posterior, 8.76% below). They are different
numbers, sometimes by a factor of ten, and the second one is the only one the
patient cares about. Bayes' theorem is the rule that converts between them, and
it is the reason a Unit 5 medical classifier reports precision and recall rather
than a single accuracy figure.


# Probabilistic Reasoning: Bayes' Theorem

**Unit:** Unit 2: AI Concepts, Terminology, and Application Domains  

## 📚 Learning Objectives

By completing this notebook, you will:
- Understand uncertainty in AI and why it matters
- Learn Bayes' theorem and its formula
- Apply Bayes' rule to real-world problems
- Implement Bayesian reasoning in Python
- Understand Bayesian statistics applications

---

## Bayes' Theorem Formula

**P(A|B) = P(B|A) × P(A) / P(B)**

Where:
- **P(A|B)**: Posterior probability (probability of A given B)
- **P(B|A)**: Likelihood (probability of B given A)
- **P(A)**: Prior probability (probability of A)
- **P(B)**: Evidence (probability of B)

**In AI Context:**
- A = Hypothesis (e.g., patient has disease)
- B = Evidence (e.g., test result)
- P(A|B) = Updated belief after seeing evidence


## 📥 Inputs & 📤 Outputs

**Inputs:** What we use in this notebook

- Libraries and concepts as introduced in this notebook; see prerequisites and code comments.

**Outputs:** What you'll see when you run the cells

- Printed results, figures, and summaries as shown when you run the cells.

---

In [1]:
# Example 1: Medical Diagnosis with Bayes' Theorem

def bayes_theorem(prior, likelihood, evidence):
    """
    Calculate posterior probability using Bayes' theorem
    
    Args:
        prior: P(A) - prior probability
        likelihood: P(B|A) - likelihood
        evidence: P(B) - evidence probability
    
    Returns:
        P(A|B) - posterior probability
    """
    posterior = (likelihood * prior) / evidence
    return posterior

# Medical test example
# Suppose:
# - Disease prevalence (prior): 1% of population
# - Test sensitivity (likelihood): 95% (P(positive|disease))
# - Test specificity: 90% (P(negative|no_disease))
# - What is P(disease|positive_test)?

prior_disease = 0.01  # 1% have the disease
sensitivity = 0.95    # 95% true positive rate
specificity = 0.90    # 90% true negative rate

# Calculate P(positive_test)
# P(positive) = P(positive|disease) * P(disease) + P(positive|no_disease) * P(no_disease)
p_positive_given_disease = sensitivity
p_positive_given_no_disease = 1 - specificity  # False positive rate
p_no_disease = 1 - prior_disease

p_positive = (p_positive_given_disease * prior_disease) + \
             (p_positive_given_no_disease * p_no_disease)

# Calculate posterior: P(disease|positive_test)
p_disease_given_positive = bayes_theorem(
    prior=prior_disease, likelihood=p_positive_given_disease,
    evidence=p_positive
)

print("=== Medical Diagnosis Example ===")
print(f"Prior probability of disease: {prior_disease:.2%}")
print(f"Test sensitivity: {sensitivity:.2%}")
print(f"Test specificity: {specificity:.2%}")
print(f"\nProbability of positive test: {p_positive:.2%}")
print(f"\nPosterior probability (disease given positive test): {p_disease_given_positive:.2%}")
print(f"\nEven with a positive test, probability is only {p_disease_given_positive:.2%}!")

=== Medical Diagnosis Example ===
Prior probability of disease: 1.00%
Test sensitivity: 95.00%
Test specificity: 90.00%

Probability of positive test: 10.85%

Posterior probability (disease given positive test): 8.76%

Even with a positive test, probability is only 8.76%!


In [2]:
# Example 2: Spam Email Detection

def spam_detector(word, prior_spam, word_given_spam, word_given_ham):
    """
    Simple spam detector using Bayes' theorem
    """
    # P(spam|word) = P(word|spam) * P(spam) / P(word)
    # P(word) = P(word|spam) * P(spam) + P(word|ham) * P(ham)
    
    prior_ham = 1 - prior_spam
    p_word = (word_given_spam * prior_spam) + (word_given_ham * prior_ham)
    
    # Calculate P(spam|word) using Bayes' theorem
    p_spam_given_word = (word_given_spam * prior_spam) / p_word
    return p_spam_given_word

# Example: Word "free" in email
prior_spam_prob = 0.3  # 30% of emails are spam
p_free_given_spam = 0.8  # 80% of spam emails contain "free"
p_free_given_ham = 0.1    # 10% of ham emails contain "free"

spam_probability = spam_detector(
    word="free", prior_spam=prior_spam_prob,
    word_given_spam=p_free_given_spam,
    word_given_ham=p_free_given_ham
)

print("\n=== Spam Detection Example ===")
print(f"Prior spam probability: {prior_spam_prob:.2%}")
print(f"P('free'|spam): {p_free_given_spam:.2%}")
print(f"P('free'|ham): {p_free_given_ham:.2%}")
print(f"\nP(spam|'free'): {spam_probability:.2%}")
print(f"\nEmail with 'free' is {spam_probability:.2%} likely to be spam")


=== Spam Detection Example ===
Prior spam probability: 30.00%
P('free'|spam): 80.00%
P('free'|ham): 10.00%

P(spam|'free'): 77.42%

Email with 'free' is 77.42% likely to be spam


## 📊 One theorem, two priors, two very different answers

The two cells applied the identical rule and changed one input — the **prior**:

| example | prior P(H) | evidence strength | posterior P(H \| E) | belief moved |
|---|---|---|---|---|
| Disease test | **1%** | sensitivity 95%, specificity 90% | **8.76%** | ×8.8, still small |
| Spam word "free" | **30%** | P(free\|spam) 80% vs P(free\|ham) 10% | **77.42%** | ×2.6, now decisive |

**The conclusion these numbers support:** the *same* strength of evidence produces
a confident answer in one case and an inconclusive one in the other, and the only
difference is the base rate. The disease test is the stronger instrument on paper
— 95% sensitivity against the spam word's 80% — and it still leaves you at 8.76%,
because it is looking for something that is rare. **Evidence updates belief; it
does not replace it.**

**The practical rule to take away:** whenever someone quotes a detection rate, ask
what fraction of the population has the thing being detected. Without that number,
the detection rate alone cannot tell you what a positive result means.

### 🔁 The same reasoning shows up here too

- **Airport and border screening:** a highly accurate check applied to a
  population where the target is very rare produces overwhelmingly false alerts.
- **Bank fraud alerts:** fraud is a fraction of a percent of transactions, which
  is why fraud teams are judged on precision, not on detection rate.
- **Rare-disease genetic testing:** the reason clinical guidelines restrict some
  tests to high-risk groups is that raising the prior is the only way to make the
  posterior usable.


## 💬 Discuss

Read the two printed results before answering: the disease test moved belief from
a 1% prior to an **8.76% posterior**, and the word "free" moved a 30% prior to
**77.42%**.

1. A positive test raised the probability of disease from 1% to 8.76% — so about
   **91 of every 100 people who test positive do not have the disease**. Would you
   run this screening programme? Argue it as the health ministry paying for the
   follow-up tests, then argue it as one of the 1% who has the disease. Nothing
   about the arithmetic changes; say what does.
2. The spam example used a single word. Real filters use thousands and assume the
   words are independent given the class — which is plainly false (*free* and
   *offer* travel together). The assumption is wrong and the filters work anyway.
   **Why might a wrong assumption still produce good decisions**, and what kind of
   error should you expect it to produce when it fails?
3. **Regional practice:** a diagnostic model is trained where a disease affects 1
   in 100, and deployed where it affects 1 in 2,000. The likelihood terms are
   unchanged. Work out what happens to the posterior, then decide what you would
   demand from a vendor before letting the model into a Saudi hospital.


## ⚠️ Where this breaks

Bayes' theorem is arithmetic — it is always correct. What breaks is everything
you have to supply to use it.

- **The prior has to come from somewhere.** We typed `0.01`. In practice that
  number comes from an epidemiological study on some population, at some time, and
  it does not transfer: prevalence differs by country, age band, season and how
  people were selected into the sample. **A posterior is only as defensible as
  the prior underneath it**, and priors are where the arguing happens.
- **Sensitivity and specificity are estimates too.** 95% and 90% were measured on
  a validation cohort with its own selection effects. If the test performs
  differently on the population you deploy to, the posterior is wrong and nothing
  in the calculation warns you.
- **The spam calculation assumes conditional independence, which is false.**
  Naive Bayes multiplies per-word probabilities as if the words were unrelated.
  Because they are correlated, it typically produces **badly calibrated
  confidences** — 0.99 when it should say 0.80 — even when the *ranking* it
  produces is good enough to classify with. Use it to decide; do not quote its
  probability as a probability.
- **One piece of evidence is the easy case.** Combining several tests requires
  either an independence assumption (usually wrong) or a full joint model, and the
  number of parameters explodes with the number of variables. That is what
  Bayesian networks (Course 02, Unit 3) exist to manage.
- **Base rates move.** Prevalence changes with an outbreak, a screening campaign,
  or a change in who gets referred. A system that hard-codes a prior degrades
  silently as the world shifts underneath it — and the output looks exactly as
  confident as it did on day one.
- **Cheaper alternative when you cannot defend a prior:** report the likelihood
  ratio instead (`P(evidence | disease) / P(evidence | no disease)`), and let the
  clinician supply the prior they believe. That separates what your data measured
  from what someone assumed.


## 📚 References

1. Bayes, T. (1763). *An Essay towards Solving a Problem in the Doctrine of Chances*. Philosophical Transactions of the Royal Society, 53, 370–418.
2. Pearl, J. (1988). *Probabilistic Reasoning in Intelligent Systems: Networks of Plausible Inference*. Morgan Kaufmann.
3. Bishop, C. M. (2006). *Pattern Recognition and Machine Learning*, Chs. 1–2. Springer.
4. Papamarkou, T., Skoularidou, M., Palla, K., et al. (2024). *Position: Bayesian Deep Learning is Needed in the Age of Large-Scale AI*. arXiv. <https://arxiv.org/abs/2402.00809>